generate/templates/manual/src/lock_master.cc

```diff
@@ -0,0 +1,231 @@
+#include <git2.h>
+#include <uv.h>
+#include <set>
+#include <vector>
+#include <map>
+#include <algorithm>
+
+#include "../include/lock_master.h"
+
+// information about a lockable object
+// - the mutex used to lock it and the number of outstanding locks
+struct ObjectInfo {
+  uv_mutex_t *mutex;
+  unsigned useCount;
+
+  ObjectInfo(uv_mutex_t *mutex, unsigned useCount)
+    : mutex(mutex), useCount(useCount)
+  {}
+};
+
+// LockMaster implementation details
+// implemented in a separate class to keep LockMaster opaque
+class LockMasterImpl {
+  // STATIC variables / methods
+
+  // A map from objects that are locked (or were locked), to information on their mutex
+  static std::map<const void *, ObjectInfo> mutexes;
+  // A mutex used for the mutexes map
+  static uv_mutex_t mapMutex;
+
+  // A libuv key used to store the current thread-specific LockMasterImpl instance
+  static uv_key_t currentLockMasterKey;
+
+  // A libuv async handle used to trigger / debounce mutex cleanup
+  static uv_async_t cleanupMutexesHandle;
+  // Cleans up any mutexes that are not currently used
+  static void CleanupMutexes(uv_async_t *async);
+
+public:
+  static void Initialize();
+
+  // INSTANCE variables / methods
+
+private:
+  // The set of objects this LockMaster is responsible for locking
+  std::set<const void *> objectsToLock;
+
+  // Mutexes locked by this LockMaster on construction and unlocked on destruction
+  std::vector<uv_mutex_t *> GetMutexes(int useCountDelta);
+  void Register();
+  void Unregister();
+  void CleanupMutexes();
+
+public:
+  static LockMasterImpl *CurrentLockMasterImpl() {
+    return (LockMasterImpl *)uv_key_get(&currentLockMasterKey);
+  }
+
+  LockMasterImpl() {
+    Register();
+    Lock(true);
+  }
+
+  ~LockMasterImpl() {
+    Unregister();
+    Unlock(true);
+    CleanupMutexes();
+  }
+
+  void ObjectToLock(const void *objectToLock) {
+    objectsToLock.insert(objectToLock);
+  }
+
+  void Lock(bool acquireMutexes);
+  void Unlock(bool releaseMutexes);
+};
+
+std::map<const void *, ObjectInfo> LockMasterImpl::mutexes;
+uv_mutex_t LockMasterImpl::mapMutex;
+uv_key_t LockMasterImpl::currentLockMasterKey;
+uv_async_t LockMasterImpl::cleanupMutexesHandle;
+
+
+void LockMasterImpl::Initialize() {
+  uv_mutex_init(&mapMutex);
+  uv_key_create(&currentLockMasterKey);
+  uv_async_init(uv_default_loop(), &cleanupMutexesHandle, CleanupMutexes);
+}
+
+void LockMasterImpl::CleanupMutexes(uv_async_t *async) {
+  uv_mutex_lock(&mapMutex);
+
+  for (auto it = mutexes.begin(); it != mutexes.end(); )
+  {
+    uv_mutex_t *mutex = it->second.mutex;
+    unsigned useCount = it->second.useCount;
+    // if the mutex is not used by any LockMasters,
+    // we can destroy it
+    if (!useCount) {
+      uv_mutex_destroy(mutex);
+      free(mutex);
+      auto to_erase = it;
+      it++;
```

Comment:
I almost did that once too, going back through the code.  Then I remembered that the following line (`mutexes.erase(to_erase);`) invalidates everything referring to the erased element, including `it` before it gets incremented.  So `it` has to be incremented before we `erase`.  In `c++11`, `erase` returns an iterator that refers to the next element and that avoids the problem, but that particular functionality isn't available on some of our targeted environments.